File: app/src/main/java/dev/mike/couchtour/LocalPlaylist.kt
package dev.mike.couchtour

import androidx.media3.common.MediaItem
import androidx.room.Dao
import androidx.room.Entity
import androidx.room.ForeignKey
import androidx.room.Index
import androidx.room.Insert
import androidx.room.PrimaryKey
import androidx.room.Query
import androidx.room.Transaction
import kotlinx.coroutines.async
import kotlinx.coroutines.awaitAll
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.flow.Flow

/**
 * Local playlists spanning both backends (#12) — the same account-free shape #11's
 * [LikedTracks] was expected to reuse, except a playlist is ordered and each entry needs
 * enough context to be resolved back into a [android.media3.common.MediaItem] later (a bare
 * track id, unlike [LikedTracks], isn't enough — see D161). Room, not `SharedPreferences`:
 * this is relational (one playlist, many ordered tracks), where [Favorites]/[LikedTracks]'s
 * flat `Set<String>` doesn't fit.
 */
@Entity(tableName = "local_playlists")
data class LocalPlaylistEntity(
    @PrimaryKey val id: String,
    val name: String,
    /** Denormalized so the list screen renders without a join. */
    val trackCount: Int = 0,
    val createdAt: Long,
    val updatedAt: Long,
)

/**
 * One playlist entry. Enough to independently refetch the track at play time (D161):
 * [backend] + [trackId] + [showDate] locate a phish.in track inside its show; Relisten also
 * needs [artistSlug] (there's no fetch-by-id endpoint) and, since a show can have several
 * tapes with different track splits, [recordingId] — null falls back to the default tape,
 * same as [MusicSource.show].
 *
 * The rest ([title], [durationMs], [venueName], [artUrl]) is denormalized display data, so
 * the playlist screen renders without re-fetching every track — the same tradeoff
 * [Progress] makes for its own display fields.
 */
@Entity(
    tableName = "local_playlist_tracks",
    foreignKeys = [ForeignKey(
        entity = LocalPlaylistEntity::class,
        parentColumns = ["id"],
        childColumns = ["playlistId"],
        onDelete = ForeignKey.CASCADE,
    )],
    indices = [Index("playlistId")],
)
data class LocalPlaylistTrackEntity(
    @PrimaryKey(autoGenerate = true) val rowId: Long = 0,
    val playlistId: String,
    val position: Int,
    /** [Backend.id] — "phishin" or "relisten". */
    val backend: String,
    /** phish.in `Track.id.toString()`, or Relisten's track uuid ([PlayableTrack.id]). */
    val trackId: String,
    val showDate: String,
    val artistSlug: String? = null,
    val recordingId: String? = null,
    val title: String,
    val durationMs: Long = 0,
    val venueName: String? = null,
    val artUrl: String? = null,
)

@Dao
interface LocalPlaylistDao {
    @Query("SELECT * FROM local_playlists ORDER BY updatedAt DESC")
    fun playlists(): Flow<List<LocalPlaylistEntity>>

    @Query("SELECT * FROM local_playlists WHERE id = :id")
    suspend fun playlist(id: String): LocalPlaylistEntity?

    @Query("SELECT * FROM local_playlist_tracks WHERE playlistId = :id ORDER BY position")
    fun tracks(id: String): Flow<List<LocalPlaylistTrackEntity>>

    @Query("SELECT * FROM local_playlist_tracks WHERE playlistId = :id ORDER BY position")
    suspend fun tracksOnce(id: String): List<LocalPlaylistTrackEntity>

    @Insert
    suspend fun insertPlaylist(playlist: LocalPlaylistEntity)

    @Insert
    suspend fun insertTrack(track: LocalPlaylistTrackEntity)

    @Query("SELECT COALESCE(MAX(position), -1) FROM local_playlist_tracks WHERE playlistId = :playlistId")
    suspend fun maxPosition(playlistId: String): Int

    @Query("UPDATE local_playlists SET trackCount = trackCount + 1, updatedAt = :now WHERE id = :id")
    suspend fun incrementTrackCount(id: String, now: Long)

    @Query("DELETE FROM local_playlist_tracks WHERE rowId = :rowId")
    suspend fun deleteTrackRow(rowId: Long)

    @Query("UPDATE local_playlists SET trackCount = trackCount - 1, updatedAt = :now WHERE id = :id")
    suspend fun decrementTrackCount(id: String, now: Long)

    @Query("DELETE FROM local_playlists WHERE id = :id")
    suspend fun deletePlaylist(id: String)

    @Query("UPDATE local_playlists SET name = :name, updatedAt = :now WHERE id = :id")
    suspend fun renamePlaylist(id: String, name: String, now: Long)

    @Query("UPDATE local_playlist_tracks SET position = :position WHERE rowId = :rowId AND playlistId = :playlistId")
    suspend fun updateTrackPosition(rowId: Long, playlistId: String, position: Int)

    /** Appends [track] to the end of its playlist and bumps [LocalPlaylistEntity.trackCount]/`updatedAt`. */
    @Transaction
    suspend fun addTrack(track: LocalPlaylistTrackEntity, now: Long) {
        val position = maxPosition(track.playlistId) + 1
        insertTrack(track.copy(position = position))
        incrementTrackCount(track.playlistId, now)
    }

    /** The inverse of [addTrack] — [rowId] is [LocalPlaylistTrackEntity.rowId]. */
    @Transaction
    suspend fun removeTrack(rowId: Long, playlistId: String, now: Long) {
        deleteTrackRow(rowId)
        decrementTrackCount(playlistId, now)
    }

    @Query("UPDATE local_playlists SET updatedAt = :now WHERE id = :id")
    suspend fun touchPlaylist(id: String, now: Long)

    /** Updates positions for [orderedRowIds] sequentially and bumps playlist `updatedAt`. */
    @Transaction
    suspend fun reorderTracks(playlistId: String, orderedRowIds: List<Long>, now: Long) {
        orderedRowIds.forEachIndexed { index, rowId ->
            updateTrackPosition(rowId, playlistId, index)
        }
        touchPlaylist(playlistId, now)
    }
}

/**
 * Builds a local playlist's queue from scratch — shared by [PlayerViewModel] (phone) and
 * [PlaybackService]'s Auto browse tree, the same way [recordingTrackItems] is, so a playlist
 * queued from either place resolves identically.
 */
internal suspend fun localPlaylistQueueItems(dao: LocalPlaylistDao, playlistId: String): List<MediaItem> {
    val playlist = dao.playlist(playlistId) ?: return emptyList()
    val resolved = resolveLocalPlaylistTracks(dao.tracksOnce(playlistId))
    return localPlaylistTrackItems(playlistId, playlist.name, resolved)
}

/**
 * Refetches every stored reference into something playable — there's no fetch-by-id endpoint
 * on either backend, so this mirrors [PlayerViewModel.playTrack]/`.resume()`'s existing
 * pattern (fetch the show, find the track inside it), batched to one fetch per distinct show
 * or tape rather than one per track. A reference that no longer resolves — a deleted show, a
 * track dropped from a tape — is skipped rather than failing the whole playlist.
 *
 * The per-show/per-tape fetches run concurrently (`async` + `awaitAll`), not one at a time —
 * a playlist spanning N distinct shows used to pay N sequential round trips before playback
 * could start, which is where a 30+ second resume on a "favorites across years" mixtape came
 * from (issue: slow Android playback resume).
 */
internal suspend fun resolveLocalPlaylistTracks(refs: List<LocalPlaylistTrackEntity>): List<ResolvedLocalTrack> = coroutineScope {
    val phishShows = refs.filter { it.backend == Backend.PHISHIN.id }
        .map { it.showDate }.distinct()
        .map { date -> async { date to runCatching { PhishInApi.show(date) }.getOrNull() } }
        .awaitAll().toMap()

    val relistenRefs = refs.filter { it.backend == Backend.RELISTEN.id }
    val relistenArtists = if (relistenRefs.isEmpty()) emptyList()
        else runCatching { RelistenCatalogSource.artists() }.getOrDefault(emptyList())
    val relistenShows = relistenRefs
        .mapNotNull { ref -> ref.artistSlug?.let { Triple(it, ref.showDate, ref.recordingId) } }
        .distinct()
        .map { key ->
            async {
                val (slug, date, recordingId) = key
                val artist = relistenArtists.firstOrNull { it.id == slug }
                key to artist?.let { runCatching { RelistenCatalogSource.show(it, date, recordingId) }.getOrNull() }
            }
        }
        .awaitAll().toMap()

    refs.mapNotNull { ref ->
        when (ref.backend) {
            Backend.PHISHIN.id -> {
                val show = phishShows[ref.showDate] ?: return@mapNotNull null
                val track = show.tracks.firstOrNull { it.id.toString() == ref.trackId } ?: return@mapNotNull null
                ResolvedLocalTrack(
                    id = track.id.toString(),
                    title = track.title,
                    url = track.mp3Url.orEmpty(),
                    waveformUrl = track.waveformImageUrl,
                    // Not track.showDate/venueName — both are null on a track nested inside
                    // a show fetch (Api.kt), only populated on flatter shapes like search
                    // results and playlist entries.
                    showDate = show.date,
                    venueName = show.venueName,
                    artUrl = show.albumCoverUrl ?: show.coverArtUrls?.medium,
                    artistName = "Phish",
                    backend = Backend.PHISHIN.id,
                    likedByUser = track.likedByUser,
                    likesCount = track.likesCount,
                )
            }
            Backend.RELISTEN.id -> {
                val slug = ref.artistSlug ?: return@mapNotNull null
                val detail = relistenShows[Triple(slug, ref.showDate, ref.recordingId)] ?: return@mapNotNull null
                val track = detail.tracks.firstOrNull { it.id == ref.trackId } ?: return@mapNotNull null
                ResolvedLocalTrack(
                    id = track.id,
                    title = track.title,
                    url = track.url,
                    waveformUrl = track.waveformUrl,
                    showDate = track.showDate,
                    venueName = track.venueName,
                    artUrl = track.artUrl,
                    artistName = detail.summary.artist.name,
                    backend = Backend.RELISTEN.id,
                    flacUrl = track.flacUrl,
                )
            }
            else -> null
        }
    }
}

File: macos/Packages/CouchTourKit/Sources/CouchTourKit/LocalPlaylist.swift
import Foundation
import GRDB

// Local playlists (#59, port of Android's LocalPlaylist.kt + Progress.kt's MIGRATION_7_8).
// Account-free, spans both backends: an ordered list of refs (backend, show date, and for
// Relisten an artist slug and tape id) resolved back to real tracks at play time, since
// neither backend has a fetch-track-by-id endpoint (D161 in DECISIONS.md). New tables only,
// on the same phishin.db/dbQueue ProgressStore already opens — sharing its one connection
// rather than a second one to the same file, see ProgressStore.dbQueue.

public struct LocalPlaylist: Codable, Hashable, FetchableRecord, PersistableRecord {
    public static let databaseTableName = "local_playlists"

    public var id: String
    public var name: String
    /// Denormalized, bumped alongside every add/remove, so the playlists list renders a
    /// count with no second query per row.
    public var trackCount: Int
    public var createdAt: Int64
    public var updatedAt: Int64

    public init(id: String, name: String, trackCount: Int = 0, createdAt: Int64, updatedAt: Int64) {
        self.id = id
        self.name = name
        self.trackCount = trackCount
        self.createdAt = createdAt
        self.updatedAt = updatedAt
    }
}

/// One entry in a playlist. `trackId`/`showDate`/`artistSlug`/`recordingId` are what
/// `resolveLocalPlaylistTracks` re-fetches with; `title`/`durationMs`/`venueName`/`artUrl`
/// are denormalized purely for rendering the playlist's own list without a network call —
/// playback always uses the freshly-fetched track, never these display fields, since a
/// track's `mp3Url` isn't stored here at all.
public struct LocalPlaylistTrack: Codable, Equatable, FetchableRecord, MutablePersistableRecord {
    public static let databaseTableName = "local_playlist_tracks"

    public var rowId: Int64?
    public var playlistId: String
    public var position: Int
    public var backend: String
    public var trackId: String
    public var showDate: String
    /// Relisten only.
    public var artistSlug: String?
    /// Relisten only; nil means the tape that was the default at add-time.
    public var recordingId: String?
    public var title: String
    public var durationMs: Int64
    public var venueName: String?
    public var artUrl: String?

    public init(
        rowId: Int64? = nil, playlistId: String, position: Int = 0, backend: String, trackId: String,
        showDate: String, artistSlug: String? = nil, recordingId: String? = nil, title: String,
        durationMs: Int64, venueName: String? = nil, artUrl: String? = nil
    ) {
        self.rowId = rowId
        self.playlistId = playlistId
        self.position = position
        self.backend = backend
        self.trackId = trackId
        self.showDate = showDate
        self.artistSlug = artistSlug
        self.recordingId = recordingId
        self.title = title
        self.durationMs = durationMs
        self.venueName = venueName
        self.artUrl = artUrl
    }

    public mutating func didInsert(_ inserted: InsertionSuccess) {
        rowId = inserted.rowID
    }
}

/// GRDB wrapper mirroring Android's `LocalPlaylistDao` one-for-one.
public final class LocalPlaylistStore {
    private let dbQueue: DatabaseQueue

    public init(sharing progressStore: ProgressStore) throws {
        dbQueue = progressStore.dbQueue
        try migrator.migrate(dbQueue)
    }

    /// In-memory store for tests, not sharing a `ProgressStore`.
    public static func inMemory() throws -> LocalPlaylistStore {
        try LocalPlaylistStore(dbQueue: DatabaseQueue())
    }

    private init(dbQueue: DatabaseQueue) throws {
        self.dbQueue = dbQueue
        try migrator.migrate(dbQueue)
    }

    /// New tables only — `progress` is untouched, matching Android's `MIGRATION_7_8` (#12,
    /// D161). Registered under its own name so it applies once, whichever store (this one or
    /// `ProgressStore`) happens to open the file first.
    private var migrator: DatabaseMigrator {
        var migrator = DatabaseMigrator()
        migrator.registerMigration("v8_localPlaylists") { db in
            try db.create(table: "local_playlists") { t in
                t.column("id", .text).primaryKey()
                t.column("name", .text).notNull()
                t.column("trackCount", .integer).notNull().defaults(to: 0)
                t.column("createdAt", .integer).notNull()
                t.column("updatedAt", .integer).notNull()
            }
            try db.create(table: "local_playlist_tracks") { t in
                t.autoIncrementedPrimaryKey("rowId")
                t.column("playlistId", .text).notNull().indexed()
                    .references("local_playlists", column: "id", onDelete: .cascade)
                t.column("position", .integer).notNull()
                t.column("backend", .text).notNull()
                t.column("trackId", .text).notNull()
                t.column("showDate", .text).notNull()
                t.column("artistSlug", .text)
                t.column("recordingId", .text)
                t.column("title", .text).notNull()
                t.column("durationMs", .integer).notNull()
                t.column("venueName", .text)
                t.column("artUrl", .text)
            }
        }
        return migrator
    }

    public func playlists() throws -> [LocalPlaylist] {
        try dbQueue.read { db in
            try LocalPlaylist.order(Column("updatedAt").desc).fetchAll(db)
        }
    }

    public func playlist(id: String) throws -> LocalPlaylist? {
        try dbQueue.read { db in try LocalPlaylist.fetchOne(db, key: id) }
    }

    public func tracks(playlistId: String) throws -> [LocalPlaylistTrack] {
        try dbQueue.read { db in
            try LocalPlaylistTrack
                .filter(Column("playlistId") == playlistId)
                .order(Column("position"))
                .fetchAll(db)
        }
    }

    @discardableResult
    public func createPlaylist(name: String, now: Int64) throws -> LocalPlaylist {
        let playlist = LocalPlaylist(id: UUID().uuidString, name: name, createdAt: now, updatedAt: now)
        try dbQueue.write { db in try playlist.insert(db) }
        return playlist
    }

    public func deletePlaylist(id: String) throws {
        try dbQueue.write { db in _ = try LocalPlaylist.deleteOne(db, key: id) }
    }

    public func renamePlaylist(id: String, name: String, now: Int64) throws {
        try dbQueue.write { db in
            try db.execute(
                sql: "UPDATE local_playlists SET name = ?, updatedAt = ? WHERE id = ?",
                arguments: [name, now, id]
            )
        }
    }

    /// Appends at the end.
    public func addTrack(_ track: LocalPlaylistTrack, toPlaylist playlistId: String, now: Int64) throws {
        try dbQueue.write { db in
            let maxPosition = try Int.fetchOne(
                db, sql: "SELECT MAX(position) FROM local_playlist_tracks WHERE playlistId = ?", arguments: [playlistId]
            ) ?? -1
            var row = track
            row.rowId = nil
            row.position = maxPosition + 1
            try row.insert(db)
            try db.execute(
                sql: "UPDATE local_playlists SET trackCount = trackCount + 1, updatedAt = ? WHERE id = ?",
                arguments: [now, playlistId]
            )
        }
    }

    public func removeTrack(rowId: Int64, fromPlaylist playlistId: String, now: Int64) throws {
        try dbQueue.write { db in
            _ = try LocalPlaylistTrack.deleteOne(db, key: rowId)
            try db.execute(
                sql: "UPDATE local_playlists SET trackCount = MAX(trackCount - 1, 0), updatedAt = ? WHERE id = ?",
                arguments: [now, playlistId]
            )
        }
    }

    /// Reorders the playlist's tracks to match `orderedRowIds` and updates `updatedAt`.
    public func reorderTracks(playlistId: String, orderedRowIds: [Int64], now: Int64) throws {
        try dbQueue.write { db in
            for (index, rowId) in orderedRowIds.enumerated() {
                try db.execute(
                    sql: "UPDATE local_playlist_tracks SET position = ? WHERE rowId = ? AND playlistId = ?",
                    arguments: [index, rowId, playlistId]
                )
            }
            try db.execute(
                sql: "UPDATE local_playlists SET updatedAt = ? WHERE id = ?",
                arguments: [now, playlistId]
            )
        }
    }
}

/// Resolves a playlist's stored refs back into playable tracks: one fetch per distinct
/// show/tape rather than per track, matching `LocalPlaylist.kt`'s
/// `resolveLocalPlaylistTracks`. A show that fails to fetch — deleted, network error, a
/// stale ref — drops just its own rows from the result rather than failing the whole
/// playlist, since neither backend has fetch-track-by-id to retry against.
public func resolveLocalPlaylistTracks(_ rows: [LocalPlaylistTrack]) async -> [PlayableTrack] {
    struct GroupKey: Hashable {
        let backend: String
        let showDate: String
        let artistSlug: String?
        let recordingId: String?
    }

    func key(for row: LocalPlaylistTrack) -> GroupKey {
        GroupKey(backend: row.backend, showDate: row.showDate, artistSlug: row.artistSlug, recordingId: row.recordingId)
    }

    var order: [GroupKey] = []
    var seen = Set<GroupKey>()
    for row in rows {
        let k = key(for: row)
        if seen.insert(k).inserted { order.append(k) }
    }

    // Concurrent, not a sequential for-loop: N distinct shows/tapes used to pay N sequential
    // HTTP round trips before playback could start, reproducing Android's 30+ second resume
    // stall on a wide mixtape playlist (D175 — the bug was found and fixed there first).
    var detailsByKey: [GroupKey: ShowDetail] = [:]
    await withTaskGroup(of: (GroupKey, ShowDetail?).self) { group in
        for k in order {
            group.addTask {
                guard let backend = Backend.from(k.backend) else { return (k, nil) }
                let artist: ArtistRef
                switch backend {
                case .phishin:
                    artist = PHISH
                case .relisten:
                    guard let slug = k.artistSlug else { return (k, nil) }
                    artist = ArtistRef(backend: .relisten, id: slug, name: "")
                }
                let detail = try? await sourceFor(backend).show(artist: artist, date: k.showDate, recordingId: k.recordingId)
                return (k, detail)
            }
        }
        for await (k, detail) in group {
            if let detail { detailsByKey[k] = detail }
        }
    }

    return rows.compactMap { row in
        detailsByKey[key(for: row)]?.tracks.first { $0.id == row.trackId }
    }
}

/// Search within a local playlist (#90). Returns matches paired with their index in the
/// original, unfiltered `tracks` array — both playback and reordering (`.onMove`,
/// `moveUp`/`moveDown` in `LocalPlaylistView`) key off that original position, so the index
/// has to survive filtering intact rather than being recomputed from the filtered list's own
/// position. Port of `Catalog.kt`'s `filterByTitleIndexed`.
public func filterByTitleIndexed(_ tracks: [LocalPlaylistTrack], query: String) -> [(offset: Int, element: LocalPlaylistTrack)] {
    let trimmed = query.trimmingCharacters(in: .whitespacesAndNewlines)
    return tracks.enumerated().filter { _, track in
        trimmed.isEmpty || track.title.range(of: trimmed, options: .caseInsensitive) != nil
    }
}

extension Array where Element == LocalPlaylistTrack {
    public func filterByTitleIndexed(_ query: String) -> [(offset: Int, element: LocalPlaylistTrack)] {
        CouchTourKit.filterByTitleIndexed(self, query: query)
    }
}
